DTM-13285 Show native JS errors details.

dompuiu opened the pull request:
`jwt-auth` throws native errors or objects containing the `error_description` property for errors thrown by the API. We need to capture both types.

dompuiu committed:
DTM-13285 Show native JS errors details.

Aaronius commented:
For the errors thrown by their API, why doesn't their `toString` output whatever their `error_description` is? It sounds like something they might need to fix on their end.

dompuiu commented:
The APIs will return a JSON that once it is parsed it will be transformed in an object. If we use `${e}`, the message will look like `Error retrieving access token. [object Object]`.

Aaronius commented:
My take is they shouldn't be throwing the API result directly, but instead should be throwing something with a proper `toString` that provides a message. Anyway, can you write a comment in the code that explains why `${e.error_description || e}` isn't just `${e}` for the next engineer that comes along?

dompuiu commented:
That is why I have written a special test for this case. If you put `${e}`, it will throw an error.

dompuiu commented:
Opened an issue on `jwt-auth`: https://github.com/adobe/jwt-auth/issues/8.

Aaronius commented:
Me too https://github.com/adobe/jwt-auth/issues/9. 🤷‍♂

Aaronius commented:
This PR needs to get updated with the latest changes from jwt-auth.

Commit:
DTM-13285 Fix error logic in getIntegrationAccessToken to match how errors are returned

in @adobe/jwt-authv0.3.0

dompuiu commented on bin/__tests__/getIntegrationAccessToken.test.js:
```diff
@@ -17,6 +17,31 @@ const METASCOPES = [
 ];
 
+const getEnvConfig = o => {
+  return Object.assign(
```
Not a big deal, but here you could get rid of the return statement and the brackets.

dompuiu commented on bin/__tests__/getIntegrationAccessToken.test.js:
```diff
@@ -17,6 +17,31 @@ const METASCOPES = [
+};
+
+const getArguments = o => {
+  return Object.assign(
```
Same like above.

dompuiu commented on bin/getIntegrationAccessToken.js:
```diff
@@ -109,9 +109,14 @@ module.exports = async (
       return response.access_token;
     } catch (e) {
-      if (e.error !== 'invalid_scope' || i === METASCOPES.length - 1) {
+      const { message: errorMessage = '' } = e;
```
Would it maybe be better to have a default message instead of an empty string? Maybe `An unknown error has occurred`.?

dompuiu commented:
Since I initiated this PR, it seems I cannot submit a review with approve. Probably @Aaronius can.

Aaronius commented on bin/__tests__/getIntegrationAccessToken.test.js:
```diff
@@ -232,10 +217,33 @@ describe('getIntegrationAccessToken', () => {
+      // we bailed after the first call because it wasn't a scoping error
+      expect(mockAuth.calls.count()).toBe(1)
+
+      // This tests that if all metascopes fail, the error from the last attempt is ultimately thrown.
```
This comment doesn't apply to this test, right?

brenthosie replied:
Good catch

Commit:
DTM-13285 Provide a fallback error message if @adobe/jwt-auth throws with no e.message.

Remove comment in test file that was incorrectly documenting what the test was doing.

Aaronius commented:
Thanks @brenthosie!

Aaronius commented:
We have the same issue in the releaser (https://github.com/adobe/reactor-releaser/blob/master/bin/getIntegrationAccessToken.js#L57). @brenthosie do you want to take that as well?

Aaronius commented:
Published in v4.1.1.
